pkg/apis/portworx/v1beta1/types.go

```diff
@@ -0,0 +1,191 @@
+
+// Placement encapsulates the various kubernetes options that control where pods are scheduled and executed.
+type Placement struct {
+	meta.TypeMeta   `json:",inline"`
+	NodeAffinity    *v1.NodeAffinity    `json:"nodeAffinity,omitempty"`
+	PodAffinity     *v1.PodAffinity     `json:"podAffinity,omitempty"`
+	PodAntiAffinity *v1.PodAntiAffinity `json:"podAntiAffinity,omitempty"`
+	Tolerations     []v1.Toleration     `json:"tolerations,omitemtpy"`
+}
+
+// +genclient
+// +genclient:noStatus
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// VolumePlacementStrategy specifies a spec for volume placement in the cluster
+type VolumePlacementStrategy struct {
+	meta.TypeMeta   `json:",inline"`
+	meta.ObjectMeta `json:"metadata,omitempty"`
+	Spec            VolumePlacementSpec `json:"spec"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// VolumePlacementStrategyList is a list of VolumePlacementStrategy objects
+type VolumePlacementStrategyList struct {
+	meta.TypeMeta `json:",inline"`
+	meta.ListMeta `json:"metadata,omitempty"`
+	// Items are the list of volume placements strategy items
+	Items []VolumePlacementStrategy `json:"items"`
+}
+
+// VolumePlacementSpec specifies a set of rules for volume placement in the cluster
+type VolumePlacementSpec struct {
+	// Rules defines a list of rules as part of the placement spec. All the rules specified will
+	// be applied for volume placement.
+	// Rules that have enforcement as "required" are strictly enforced while "preferred" are best effort.
+	// In situations, where 2 or more rules conflict, the weight of the rules will dictate which wins.
+	Rules []VolumePlacementRule `json:"rules"`
+}
+
+// VolumePlacementRule defines the rule for placing volume replicas
+type VolumePlacementRule struct {
+	// AffectedReplicas defines the number of volume replicas affected by this rule. If not provided,
+	// rule would affect all replicas
+	// (optional)
+	AffectedReplicas int64 `json:"affectedReplicas,omitempty"`
+	// Weight defines the weight of the rule which allows to break the tie with other matching rules. A rule with
+	// higher weight wins over a rule with lower weight.
+	// (optional)
+	Weight int64 `json:"weight,omitempty"`
+	// Enforcement specifies the rule enforcement policy. Can take values: required or preferred.
+	// (optional)
+	Enforcement api.VolumePlacementRule_EnforcementType `json:"enforcement,omitempty"`
+	// Type is the type of the affinity rule
+	Type api.VolumePlacementRule_AffinityRuleType `json:"type,omitempty"`
+	// MatchExpressions is a list of label selector requirements. The requirements are ANDed.
+	MatchExpressions []*LabelSelectorRequirement `json:"matchExpressions,omitempty"`
+}
+
+// LabelSelectorRequirement A label selector requirement is a selector that contains values, a key, and an operator that
+// relates the key and values.
+type LabelSelectorRequirement struct {
```

Comment:
should we use the LabelSelector object from k8s instead of defining on our side?

Reply:
Done